symforce/codegen/backends/cpp/templates/tests/cam_package_cpp_test.cc.jinja

```diff
@@ -137,7 +137,7 @@ TEMPLATE_TEST_CASE("Test storage ops", "[cam_package]", {{ ", ".join(cpp_cam_typ
 
   std::array<Scalar, storage_dim> arr;
   cam_cal.ToStorage(arr.data());
-  for (int i = 0; i < arr.size(); ++i) {
+  for (int i = 0; i < static_cast<int>(arr.size()); ++i) {
```

Comment:
Why not `for (size_t i = 0; i < arr.size(); ++i) {` here? Just a style preference or is there a deeper reason?

Reply:
Yeah, copy-pasting some internal discussion here:

> Aaron:

What's the best practice for warnings like this?
```
/home/aaron/symforce/symforce/opt/[linearizer.cc](http://linearizer.cc/): In instantiation of ‘void sym::Linearizer<ScalarType>::UpdatePatternFromSparseFactorIntoTripletLists(const LinearizedSparseFactor&, const sym::linearization_sparse_factor_helper_t&, std::vector<Eigen::Triplet<ScalarType> >*, std::vector<Eigen::Triplet<ScalarType> >*) const [with ScalarType = float; sym::Linearizer<ScalarType>::LinearizedSparseFactor = sym::linearized_sparse_factorf_t; typename Eigen::SparseMatrix<Scalar>::StorageIndex = int]’:
/home/aaron/symforce/symforce/opt/linearizer.cc:550:21:   required from here
/home/aaron/symforce/symforce/opt/linearizer.cc:420:29: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   for (int key_i = 0; key_i < factor_helper.key_helpers.size(); key_i++) {
                       ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I tend to think size_t was a mistake when combined with C++ treating unsigned overflow as well-defined and unsigned overflow as UB, based on I think one of Chandler Carruth's CppCon talks (I don't remember which one, I could go looking...).  Which I think would mean the correct thing to do here to not footgun yourself in terms of performance to is something like:
```
   for (int64_t key_i = 0; key_i < static_cast<int64_t>(factor_helper.key_helpers.size()); key_i++) {
```
But maybe the more correct thing is to do:
```
   for (size_t key_i = 0; key_i < factor_helper.key_helpers.size(); key_i++) {
```
at the cost of potentially making things unnecessarily slow?

Or another alternative is to pass the compiler -Wno-sign-compare and disable the warning, which Hayk seems to agree is kinda sketchy.

The [Google Style Guide](https://google.github.io/styleguide/cppguide.html#Integer_Types) seems to indicate to try and use iterators here, but assuming it's a loop that doesn't want to use iterators, to not declare the loop variable as unsigned, which means casting the container size to unsigned?  And probably using int64_t?

Anyone know more or know if there's a best practice here?

The reason it can be slow is that unsigned overflow is defined behavior, which means that whatever instructions the compiler emits have to have the correct behavior when that overflow happens, which can prevent it from doing the most efficient thing in some cases. Signed overflow is UB, so the compiler can emit the most efficient thing without needing to do the right thing if the loop counter overflows (all according to the talk, which did have a real-world demo of this on a tight loop)

Which I believe is the reason (at least one of the reasons) why the Google style guide mentions this is considered a problem with the standard (in the unsigned integers section)

> Peter:

First of all it's a pretty tiny performance difference I'm almost certain.

I personally tend to avoid unsigned values as much as possible to avoid potential confusing wrapping behavior for "negative" numbers treated as unsigned

So I would cast the size() to a signed value to avoid the warning. This kind of goes along with google in correcting the "historical mistake"

> Aaron:

I think this is how I'm leaning too - agree it's a tiny performance gain (for most loops, iirc for the tightest loops in some cases it can be more significant), but I'd like to figure out what the "most correct" thing is here since I'd rather just pick the right thing and use that everywhere in symforce